### PR TITLE
Maximum Volume Restrictions Manager

### PR DESCRIPTION
**Pull request by vyomshm**
### Please check if the PR fulfills these requirements
- [x] The commit message follows our Submission guidelines
- [x] Tests for the changes have been added (for bug fixes / features)

### What kind of change does this PR introduce? 
feature - ref #267 

### What is the current behavior? 
#267 
No way to add global token volume transfer restrictions

### What is the new behavior?
Feature added - A new Transfer Manager module
Issuers can now create(as well as remove and modify) multiple maximum volume transfer restrictions which will be applied globally to every transfer

### Does this PR introduce a breaking change?
(What changes might users need to make in their application due to this PR?)
No

### CODE REVIEW

**Comment by satyamakgec** [comment]
Hey, @vyomshm let me know when should I start reviewing your code?

**Comment by vyomshm** [comment]
Hey @satyamakgec Please review my PR

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        RollingIntervals rollingPeriodInterval;
+    }
+
+    MaximumVolumeRestriction[] public maximumVolumeRestrictions;
+
+    // mapping from address to RollingIntervals to periodIds to tokens transfered 
+    mapping(address => mapping(uint256 => mapping(uint256 => uint256))) volumeTally;
+
+    // Emit whenever a new maximum volume restriction gets added
+    event MaximumVolumeRestrictionAdded(
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `AddNewMaximumVolumeRestriction`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        address indexed from,
+        uint256 timestamp,
+        uint256 _maximumVolume, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
+    );
+
+    // Emit whenever a new maximum volume restriction gets modified
+    event MaximumVolumeRestrictionModified(
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `ModifyMaximumVolumeRestriction`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+        uint256 timestamp,
+        uint256 idx, 
+        uint256 _maximumVolume, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
+    );
+
+    // Emit whenever a new maximum volume restriction gets removed
+    event MaximumVolumeRestrictionRemoved(
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `RemoveMaximumVolumeRestricion`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /**
+     * @notice This function returns the signature of configure function
+     */
+    function getInitFunction() public pure returns (bytes4) {
+        return bytes4(0);
+    }
+
+    /// @notice Used to verify the transfer transaction according to the rule implemented in the trnasfer managers
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: add a descriptive natspec comments

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+    * @notice allows admin to modify a maximum volume transfer restriction
+    * @param idx representing the array index of the maximum volume transfer restriction
+    * @param _maximumVolume maximum token volume that can be transfered
+    * @param _startTime timestamp representing when this maximum volume restriction starts
+    * @param _endTime timestamp representing when this maximum volume restriction ends
+    * @param _rollingPeriodInterval interval period representing one of five RollingIntervals
+    */
+    function modifyMaximumTransferRestriction(
+        uint256 idx,
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `_index`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        if(_rollingPeriodInterval == RollingIntervals.Daily){
+            period = 24 hours;
+        }else if(_rollingPeriodInterval == RollingIntervals.Weekly){
+            period = 168 hours;
+        }else if(_rollingPeriodInterval == RollingIntervals.Quarterly){
+            period = 2920 hours;
+        }else if(_rollingPeriodInterval == RollingIntervals.Biannually){
+            period = 4380 hours;
+        }else{
+            period = 8760 hours;
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: change it to  365 * 24 hours that makes easy to read.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    function verifyTransfer(address _from, address /*_to*/, uint256 _amount, bytes /* _data */, bool _isTransfer) public returns(Result) {
+        // function must only be called by the associated security token if _isTransfer == true
+        require(_isTransfer == false || msg.sender == securityToken, "Sender is not owner");
+
+        if (!paused) {
+            for(uint256 i = 0; i < maximumVolumeRestrictions.length; i++){
+                MaximumVolumeRestriction memory restriction = maximumVolumeRestrictions[i];
+                if(now >= restriction.startTime && now <= restriction.endTime){
+                    uint256 period = _getPeriod(restriction.rollingPeriodInterval);
+                    uint256 periodId = now.div(period);
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: periodId should be `(now - startTime).div(period)`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    MaximumVolumeRestriction[] public maximumVolumeRestrictions;
+
+    // mapping from address to RollingIntervals to periodIds to tokens transfered 
+    mapping(address => mapping(uint256 => mapping(uint256 => uint256))) volumeTally;
+
+    // Emit whenever a new maximum volume restriction gets added
+    event MaximumVolumeRestrictionAdded(
+        address indexed from,
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: no need of `from` address. it always be owner or a delegate

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        uint256 timestamp,
+        uint256 _maximumVolume, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
+    );
+
+    // Emit whenever a new maximum volume restriction gets modified
+    event MaximumVolumeRestrictionModified(
+        address indexed from,
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: no need of `from` address. it always be owner or a delegate

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+        uint256 timestamp,
+        uint256 idx,
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: change name `idx` to `index`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /**
+    * @notice allows admin to add multiple new maximum volume transfer restrictions
+    * @param _maximumVolumes maximum token volumes that can be transfered
+    * @param _startTimes timestamps representing when these maximum volume restrictions start
+    * @param _endTimes timestamps representing when these maximum volume restrictions end
+    * @param _rollingPeriodIntervals interval periods representing one of five RollingIntervals
+    */
+    function addMultipleMaximumVolumeRestrictions(
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `addMaxVolumeRestricionsMulti()`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    /**
+    * @notice allows admin to add new maximum volume transfer restrictions
+    * @param _maximumVolume maximum token volume that can be transfered
+    * @param _startTime timestamp representing when this maximum volume restriction starts
+    * @param _endTime timestamp representing when this maximum volume restriction ends
+    * @param _rollingPeriodInterval interval period representing one of five RollingIntervals
+    */
+    function addMaximumVolumeRestriction(
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: rename it to `addMaxVolumeRestricion`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    mapping(address => mapping(uint256 => mapping(uint256 => uint256))) volumeTally;
+
+    // Emit whenever a new maximum volume restriction gets added
+    event MaximumVolumeRestrictionAdded(
+        address indexed from,
+        uint256 timestamp,
+        uint256 _maximumVolume, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:35]: emit the `index` of the new volume restriction

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+            for(uint256 i = 0; i < maximumVolumeRestrictions.length; i++){
+                MaximumVolumeRestriction memory restriction = maximumVolumeRestrictions[i];
+                if(now >= restriction.startTime && now <= restriction.endTime){
+                    uint256 period = _getPeriod(restriction.rollingPeriodInterval);
+                    uint256 periodId = now.div(period);
+                    if(_amount.add(volumeTally[_from][uint256(restriction.rollingPeriodInterval)][periodId]) > restriction.maximumVolumeAllowed){
+                        return Result.INVALID;
+                    }else{
+                        volumeTally[_from][uint256(restriction.rollingPeriodInterval)][periodId] = _amount.add(volumeTally[_from][uint256(restriction.rollingPeriodInterval)][periodId]);
+                    }
```
Comment (satyamakgec) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:95]: Add some description on your algorithm to understand the code steps

**Review by satyamakgec — CHANGES_REQUESTED** [review]

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 6 above)
Comment (F-OBrien) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Also is the number of hours for the RollingIntervals.Quarterly correct? 365*24/4 = 2190 hours not 2920

Comment (satyamakgec) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: @FOBCrypto Thanks for pointing out. Yes, it is miscalculated but this calculation get resolved if he put the hours in factors

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 1 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: done!

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 13 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:35]: done. adding an `_index` param to be the first

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 6 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Fixed it! @FOBCrypto thanks!

**Comment by vyomshm** [comment]
@satyamakgec @FOBCrypto I've made the suggested changes, please review again.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+import "./ITransferManager.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+/**
+ * @title Transfer Manager module for capping the amount of token transfers within a period
+ */
+contract MaximumVolumeTransferManager is ITransferManager {
+	
+	using SafeMath for uint256;
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Linting.
```suggestion
    using SafeMath for uint256;
```

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
+    ) internal view {
+        require(
+            uint256(RollingIntervals.Annually) >= _rollingPeriodInterval,
+            "invalid rolling period interval - should be < 5"
+        );
+        require(_startTime == 0 || _startTime >= now, 'invalid startTime');
+        require(_endTime > 0 && _endTime > _startTime, 'invalid endTime');
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: `_endTime > 0` is not required as it will always be true when `_endTime > _startTime`.
```suggestion
        require(_endTime > _startTime, 'invalid endTime');
```

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    function _checkRestrictionParams(
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval
+    ) internal view {
+        require(
+            uint256(RollingIntervals.Annually) >= _rollingPeriodInterval,
+            "invalid rolling period interval - should be < 5"
+        );
+        require(_startTime == 0 || _startTime >= now, 'invalid startTime');
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: StartTime 0 should not be allowed. `_startTime < now` should be allowed though for editing/adding restriction that have already started.
```suggestion
        require(_startTime != 0, 'invalid startTime');
```

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+        uint256 _maximumVolume, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval) 
+    public withPerm(ADMIN) {
+        _checkRestrictionParams(_startTime, _endTime, _rollingPeriodInterval);
+
+        if(_startTime == 0) {
+            _startTime = now;
+            require(_endTime > _startTime, "end time must be greater than start time");
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: No need for this require statement.
Just move the if block above `_checkRestrictionParams()`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+        require(_maximumVolumes.length >= 1, "No restrictions provided");
+        require(
+            _maximumVolumes.length == _startTimes.length &&
+            _maximumVolumes.length == _endTimes.length &&
+            _maximumVolumes.length == _rollingPeriodIntervals.length,
+            "Input array length mismatch"
+        );
+
+        for(uint256 i = 0; i < _maximumVolumes.length; i++){
+            addMaxVolumeRestricion(
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Create an internal function and use it here as well as in `addMaxVolumeRestricion()`. Doing so will save gas as parameters are not copied into memory again for internal functions.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+        uint256 _endTime, 
+        uint256 _rollingPeriodInterval) 
+    public withPerm(ADMIN) {
+        require(_index < maximumVolumeRestrictions.length, "invalid index");
+        require(now < maximumVolumeRestrictions[_index].endTime, "restriction is no longer active");
+        _checkRestrictionParams(_startTime, _endTime, _rollingPeriodInterval);
+
+        if(_startTime == 0) {
+            _startTime = now;
+            require(_endTime > _startTime, "end time must be greater than start time");
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Remove the require statement, move the if block above `_checkRestrictionParams()`

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+            _rollingPeriodInterval
+        );
+    }
+
+    /**
+    * @notice allows admin to delete a maximum volume transfer restriction
+    * @param _index representing the array index of the maximum volume transfer restriction
+    */
+    function removeMaximumTransferRestriction(uint256 _index) public withPerm(ADMIN){
+        require (maximumVolumeRestrictions.length != 0, "no existing restrictions");
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Not required. The second statement covers this as well.
```suggestion
```

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+        require (maximumVolumeRestrictions.length > _index, "invalid index");
+
+        if(_index < (maximumVolumeRestrictions.length - 1)){
+            MaximumVolumeRestriction memory lastRestriction = maximumVolumeRestrictions[maximumVolumeRestrictions.length - 1];
+            maximumVolumeRestrictions[_index] = lastRestriction;
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: No need to copy it in memory first. Just copy directly from storage.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+        }
+
+        maximumVolumeRestrictions.length = (maximumVolumeRestrictions.length).sub(1);
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: SafeMath is not required here as it will revert before reaching this statement if `maximumVolumeRestrictions.length == 0`
```suggestion
        maximumVolumeRestrictions.length = maximumVolumeRestrictions.length - 1;
```

File: scripts/test.sh [hunk 24]
```diff
@@ -63,7 +63,7 @@ start_testrpc() {
   if ! [ -z "${TRAVIS_PULL_REQUEST+x}" ] && [ "$TRAVIS_PULL_REQUEST" != false ]; then
     node_modules/.bin/testrpc-sc --gasLimit 0xfffffffffff --port "$testrpc_port" "${accounts[@]}" > /dev/null &
   else
-    node_modules/.bin/ganache-cli --gasLimit 8000000 "${accounts[@]}" > /dev/null &
+    node_modules/.bin/ganache-cli --gasLimit 0xfffffffffff "${accounts[@]}" > /dev/null &
```
Comment (maxsam4) [inline scripts/test.sh]: We need to keep our tests within the limit of 8m gas (maximum allowed on main net)
```suggestion
    node_modules/.bin/ganache-cli --gasLimit 8000000 "${accounts[@]}" > /dev/null &
```

**Review by maxsam4 — CHANGES_REQUESTED** [review]
Other than the suggested changes, Can you please also increase the coverage of your code?
Branch coverage of `MaximumVolumeTransferManager.sol` is at 80% currently. We would ideally like to see 100% coverage but a minimum of 95% is required.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 17 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: okay but accordingly I've added `_endTime > now` to prevent anyone from modifying expired restrictions or adding invalid ones.
```suggestion
        require(_startTime == 0 || _startTime >= now, 'invalid startTime');
        require(_endTime > _startTime && _endTime > now, 'invalid endTime');
```

**Comment by maxsam4** [comment]
@vyomshm Please push the changes before marking conversations as resolved.

**Comment by vyomshm** [comment]
@satyamakgec @maxsam4 @FOBCrypto I've increased my test coverage + implemented changes as discussed.  Ready for another review👍

**Comment by maxsam4** [comment]
Nice work @vyomshm !

The branch coverage is still at 92.5% with 3 uncovered branches. I would love for them to be covered as well (all of them can be covered). You can see the uncovered branches at https://coveralls.io/builds/19635412/source?filename=contracts%2Fmodules%2FTransferManager%2FMaximumVolumeTransferManager.sol

**Comment by vyomshm** [comment]
@maxsam4 @satyamakgec @FOBCrypto Coverage is now at 100%

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 25]
```diff
@@ -134,7 +134,6 @@ contract MaximumVolumeTransferManager is ITransferManager {
         uint256[] _endTimes, 
         uint256[] _rollingPeriodIntervals) 
     public withPerm(ADMIN) {
-        require(_maximumVolumes.length >= 1, "No restrictions provided");
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: Put the sanity check back.
Add a test case with empty arrays to cover the branch.

**Review by maxsam4 — CHANGES_REQUESTED** [review]

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 25 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: done!

**Comment by vyomshm** [comment]
@satyamakgec @FOBCrypto @maxsam4 made the changes. please have a look

**Comment by vyomshm** [comment]
@maxsam4 have you had a chance to look through this?  Would love to see this get resolved before you guys get too busy with pre-devcon4 events  :)

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 26]
```diff
@@ -0,0 +1,307 @@
...
+            _endTime,
+            _rollingPeriodInterval
+        );
+    }
+
+    /**
+    * @notice allows admin to delete a maximum volume transfer restriction
+    * @param _index representing the array index of the maximum volume transfer restriction
+    */
+    function removeMaximumTransferRestriction(uint256 _index) public withPerm(ADMIN){
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:201]: Can you please add a batch/multi versions of this.

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol [hunk 27]
```diff
@@ -0,0 +1,307 @@
...
+
+    /**
+    * @notice allows admin to modify a maximum volume transfer restriction
+    * @param _index representing the array index of the maximum volume transfer restriction
+    * @param _maximumVolume maximum token volume that can be transfered
+    * @param _startTime timestamp representing when this maximum volume restriction starts
+    * @param _endTime timestamp representing when this maximum volume restriction ends
+    * @param _rollingPeriodInterval interval period representing one of five RollingIntervals
+    */
+    function modifyMaximumTransferRestriction(
```
Comment (maxsam4) [inline contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:163]: Can you please add a batch/multi versions of this.

**Review by maxsam4 — CHANGES_REQUESTED** [review]

File: contracts/modules/TransferManager/MaximumVolumeTransferManager.sol (hunk 26 above)
Comment (vyomshm) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:201]: @satyamakgec  @maxsam4 Currently, removing a single array element will almost always change the order of restrictions in the array( except when the restriction to be deleted in the last one). This could render some of the indices provided unusable and therefore cause the function to revert.

I could write an implementation which would preserve the order of restrictions by shifting the elements instead of swapping them(as implemented now) but doing so would make the function extremely costly in terms of gas. 
```
string element = myArray[index];
for (uint i = index; i < myArray.length - 1; i++) {
  myArray[index] = myArray[index + 1];
}
myArray.length--;
```

Please advise on how to proceed with this.

Comment (maxsam4) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol]: @vyomshm Can you update the indices as well when deleting an array element?

Comment (maxsam4) [reply contracts/modules/TransferManager/MaximumVolumeTransferManager.sol:201]: I think one way around the problem can be by adding a unique id to `MaximumVolumeRestriction` struct and maintaining an array of that id for parsing. `maximumVolumeRestrictions` can be made a mapping rather than an array then.

**Review by maxsam4 — CHANGES_REQUESTED** [review]
I was going through the logic once again and it seems flawed. There can be multiple restrictions of overlapping periods which can make the logic fail. For example, consider a case when there is a monthly restriction started on 1st Jan and another monthly restriction started on 15th Jan. If anyone tries to transfer tokens on 16th of Jan, the below line will be executed twice
`volumeTally[_from][monthly][0] = _amount.add(volumeTally[_from][monthly][0]);`
This should have been executed only once.

There are a lot more problematic scenarios due to this. Another example with the same restrictions will be when a transfer is done on 1st Feb.  `volumeTally[_from][monthly][0]` will now contain info about the volume of trade that happened on 1st of Jan but the second restriction will not be expecting that.

Basically, the logic will fail when there are overlapping trade restriction periods. Please, rethink about it and try to fix it.

**Comment by maxsam4** [comment]
I was just going through the requirements and specs of this module and seems like your implementation and our requirements don't completely match. Please have a look at https://www.lucidchart.com/invitations/accept/75bdef55-cd39-48d9-bc1c-ac1d2fc76a39 again.

We want different people to be linked with different volume restrictions. Not all restrictions need to be imposed on every user.

**Comment by vyomshm** [comment]
@maxsam4 With regards to your first comment, `periodId` in both cases would not be the same as it is dependent on `startTime` of a restriction  and defined as 
`uint256 periodId = (now.sub(restriction.startTime)).div(period);` 
therefore, the line 
`volumeTally[_from][uint256(restriction.rollingPeriodInterval)][periodId] = _amount.add(volumeTally[_from][uint256(restriction.rollingPeriodInterval)][periodId]);`

would be executed only when relevant.

**Comment by vyomshm** [comment]
`This applies to all token-holders and not just affiliated parties.`
@maxsam4 I'll admit that from this line in "Short summary" section of #267  as posted by @CPSTL  as well as the absence of any `address` field in the spec diagram, I **inferred that all max volume transfer restrictions were supposed to be globally imposed on all token holders and not just specific addresses**
screenshot 2018-10-27 at 12 34 39 am

Please clarify if my interpretation of the module is wrong, so I can rectify my codebase.

**Comment by maxsam4** [comment]
> @maxsam4 With regards to your first comment, `periodId` in both cases would not be the same as it is dependent on `startTime` of a restriction and defined as
> `uint256 periodId = (now.sub(restriction.startTime)).div(period);`
> would be executed only when relevant.

@vyomshm Solidity does not support decimals which means after `div(period)` we will have the same `periodId` for a lot of restrictions. All restrictions will start with periodId 0 and keep incrementing as time passes. Please let me know if I am understanding it wrong.



> Please clarify if my interpretation of the module is wrong, so I can rectify my codebase.

I agree that we could have been clearer about the specs but as the second assumption in lucidchart states, the token owner should be able to impose different restrictions on different sets of token holders. Please rework your approach accordingly. 
Thank you for cooperating.

**Comment by vyomshm** [comment]
@CPSTL  @satyamakgec @maxsam4 Just to be clear, we want maxVolume transfer restrictions to exist for specific addresses only and not as Global impositions ?

This would require changing `addMaxVolumeRestricion` function signature to also include an `address` field. 
Also I could modify the `maximumVolumeRestrictions` to be a mapping instead of a global array like -
`mapping(address => MaximumVolumeRestriction) public maximumVolumeRestrictions`. This would also resolve the problem you described.

Can you confirm this is the approach  I should proceed with? 
I believe this would lead to an implementation that is quite similar as being taken for this issue -> https://github.com/PolymathNetwork/polymath-core/issues/268 .

**Comment by maxsam4** [comment]
@vyomshm We want both the abilities. Global restrictions + restrictions on only a subset of address.
This module should cater to global restrictions and the other volume restriction module should cater to individual limits. 
The limits on a subset of investors can be catered to by both or any one of these modules. I'll have a chat with the team and get back to you regarding this.

**Comment by vyomshm** [comment]
@maxsam4 okay, I'll wait.

Just so you know, I've figured out the easiest fix for the same periodId issue. As in my original implementation we should define periodId as follows-
`uint256 periodId = now.div(period);`

Also I believe we can implement the multi version of `removeMaximumTransferRestriction` by dynamically modifying the user provided indices according to elements already removed.

**Comment by maxsam4** [comment]
> @maxsam4 okay, I'll wait.
> 
> Just so you know, I've figured out the easiest fix for the same periodId issue. As in my original implementation we should define periodId as follows-
> `uint256 periodId = now.div(period);`

That won't work for the same reasons I said when you first implemented it. Also, Even in this case, there will be collisions because solidity does not support floating points. It will round them to the nearest integer.

**Comment by vyomshm** [comment]
@maxsam4 I understand what you're saying. As I see it the best way to avoid any sort of collisions will be to maintain a separate `volumeTally` for every restriction.

One way to do this would be to add to modify the `MaximumVolumeRestriction` struct to also include an `id`  and extending the multi-level `volumeTally` mapping another level.
```
uint256 internal _restrictionIdCount = 0;

struct MaximumVolumeRestriction{
        uint256 restrictionId;
        uint256 maximumVolumeAllowed;
        uint256 startTime;
        uint256 endTime;
        RollingIntervals rollingPeriodInterval;
}

// mapping from address to RollingIntervals to restrictionIds to periodIds to tokens transfered 
mapping(address => mapping(uint256 => mapping( uint256 => mapping(uint256 => uint256))) volumeTally;
```
`restrictionId` var could take its value from a global `restrictionIdCount` which would in turn be incremented by 1 every time a maximum volume restriction is created/modified. This would ensure that all `maxVolumeRestrictions` have a unique `restrictionId`

**Comment by vyomshm** [comment]
@maxsam4 have you had a chance to talk with the team yet?

**Comment by maxsam4** [comment]
@vyomshm Sorry, we are busy with devcon this week. I'll get back to you next week. Thanks for understanding.

**Comment by vyomshm** [comment]
@maxsam4 Could you update me on how to proceed with this?

**Comment by maxsam4** [comment]
Bounty paid. Thanks a lot for your contribution @vyomshm .
